Optimize block entity removal by using a ArrayList with bulk remove by index

MrPowerGamerBR opened the pull request:
The `blockEntityTickers` list was replaced with a custom list based on fastutil's `ObjectArrayList` with a small yet huge change for us: A method that allows us to remove a list of indexes from the list.

This is WAY FASTER than using `removeAll` with a list of entries to be removed, because we don't need to calculate the identity of each block entity to be removed, and we can jump directly to where the search should begin, giving a performance boost for small removals (because we don't need to loop thru the entire list to find what element should be removed, and we can quit after removing all the indexes) and a performance boost for big removals (no need to calculate the identity of each block entity), ESPECIALLY if the server has a lot of block entities, or if you are unloading a lot of block entities.

While I don't have nice and fancy and fluffy benchmarks on hand, here's some results from a quick and dirty benchmark test...
* If you have 10k block entities and attempt to remove the first block entity...
   * Current implementation: ~6333.6 op/s
   * New implementation: ~33745.2 op/s
* If you have 10k block entities and attempt to remove the last block entity...
   * Current implementation: ~6726.4 op/s
   * New implementation: ~34487.8 op/s
* If you have 10k block entities and attempt to remove block entities from 900 to 915 (simulating what happens in a chunk unload)
   * Current implementation: ~5073.6 op/s
   * New implementation: ~31476.6 op/s
* If you have 10k block entities and attempt to remove block entities from it % 2 == 0 (this is like, the worst case scenario for us)
   * Current implementation: ~1372.2 op/s
   * New implementation: ~2895.6 op/s

**Notes:**
* The patch is a bit messy, but I can clean up after the idea is "approved".
* Yes, this is faster than if you manually used `remove(int)` for each block entity, since that would create a arraycopy for each removal, and that makes us have [the big sad](https://pbs.twimg.com/media/EAmr-PAWsAEoiWR.jpg). (see patch below, that's how Minecraft originally works and Paper replaced it with a better implementation)
* I'm not 100% sure if the `removeAllByIndex` is good enough: I have tested it a lot and run the server with both implementations (old and the new) running with `equals` checking both of them to see if the order of both lists matched, and it seems to be fine.
* The `removeAllByIndex` is somewhat based on `removeElements` from fastutil, originally I was attempting to "remove block entities by range" by using `removeElements` but that wasn't good if the ranges were scattered across the list, which can happen if multiple chunks with low block entity counts unload.
* ~~*Maybe* this patch should be removed beforehand? Or maybe that patch should be edited? They both edit the same section attempting to improve removal performance. https://github.com/PaperMC/Paper/blob/master/patches/server/0164-Fix-MC-117075-TE-Unload-Lag-Spike.patch~~
  * This patch now replaces the old patch
* I think that the list doesn't need to be named `BlockEntityTickersList`, considering that it doesn't have any specific block entity behavior, however I don't know a good name for it. `ObjectArrayListWithBulkRemovalByIndex` seems wonky.
* A super hacky patch is here if you want to compare the performance between the old behavior and the new behavior, it logs to the console what implementation was fastest to remove, requires some changes to make it work on Paper tho: https://github.com/SparklyPower/SparklyPaper/blob/b96d6f8f3895d944b3c6f52fb4e0f8d706ba3f3a/patches/removed/server/0019-BlockEntityTickersList-optimization-tests.patch#L1

MrPowerGamerBR committed:
Optimize block entity removal by using a ArrayList with bulk remove by index

The blockEntityTickers list was replaced with a custom list based on fastutil's ObjectArrayList with a small yet huge change for us: A method that allows us to remove a list of indexes from the list.

This is WAY FASTER than using removeAll with a list of entries to be removed, because we don't need to calculate the identity of each block entity to be removed, and we can jump directly to where the search should begin, giving a performance boost for small removals (because we don't need to loop thru the entire list to find what element should be removed) and a performance boost for big removals (no need to calculate the identity of each block entity).

Machine-Maker reviewed (changes requested):
This patch should just completely replace `Fix MC-117075: TE Unload Lag Spike`. That is the patch that changes the removal logic from vanilla's removal 1 at a time to using the removeAll. No point in having a patch that is completely reverted in a later patch.

MrPowerGamerBR committed:
Replace old MC-117075 lag spike patch, move start index handling to the list itself

MrPowerGamerBR committed:
Thanks for optimizing the imports, IntelliJ IDEA

MrPowerGamerBR committed:
"this"

MrPowerGamerBR committed:
Don't manually edit the patch files, worst mistake of my life

MrPowerGamerBR committed:
Use IntOpenHashSet instead of HashSet, make the class final and other small fixes

Who would've thought that using the IntOpenHashSet directly would have actually avoided autoboxing?? (crazy)

Machine-Maker reviewed (approved):
Approving as the concept and implementation seem fine. I don't know if someone else like @Spottedleaf might have a different preferred way to remove the tickers. 

I also think the final product should use better param/local names as single letters here can just be super confusing.

MrPowerGamerBR committed:
Change variable names and other small changes

kennytv commented on patches/server/0164-Fix-MC-117075-TE-Unload-Lag-Spike.patch:
```diff
@@ -1,42 +1,158 @@
-Date: Wed, 9 Aug 2017 17:51:22 -0500
+From: MrPowerGamerBR <git@mrpowergamerbr.com>
+Date: Sun, 26 Nov 2023 13:02:16 -0300
 Subject: [PATCH] Fix MC-117075: TE Unload Lag Spike
```
Not related to the changes, but the patch should be renamed to say block entities instead of TE/tile entities

kennytv commented on patches/server/0164-Fix-MC-117075-TE-Unload-Lag-Spike.patch:
```diff
@@ -1,42 +1,158 @@
++     * @param index the index of the item on the list to be marked as removed
++     */
++    public void markAsRemoved(final int index) {
++        // The block entities list always loop starting from 0, so we only need to check if the startSearchFromIndex is -1 and that's it
```
While true, I don't like assuming this over code that can change or be extended at any time and possibly go by unnoticed, and adding the extra check isn't expensive, same for a bounds check

kennytv commented on patches/server/0164-Fix-MC-117075-TE-Unload-Lag-Spike.patch:
```diff
@@ -1,42 +1,158 @@
++     * @param startSearchFromIndex from where the search should start
++     * @param indexesToBeRemoved list of indexes to be removed
++     */
++    private void removeAllByIndex(final int startSearchFromIndex, final IntOpenHashSet indexesToBeRemoved) { // can't use Set<Integer> because we want to avoid autoboxing when using contains
```
IntSet

MrPowerGamerBR committed:
Rename patch, always compare if the new index is smaller than the old index, use IntSet
